mplc/multi_partner_learning/fast_mpl.py

```diff
@@ -575,3 +575,95 @@ def fit_epoch(model, train_dataset, partners_grads, smodel_list, global_grad, ag
+    and performs a optimization step with this aggregated gradient. FedGDO stands for Federated Gradient with double
+    optimizers.
+    """
+    name = 'FastFedGDO'
+
+    def __init__(self, scenario, reset_local_optims=False, global_optimiser=None, **kwargs):
+        self.global_optimiser = global_optimiser
+        self.reset_local_optims = reset_local_optims
+        super(FastFedGDO, self).__init__(scenario, **kwargs)
+
+    def init_specific_tf_variable(self):
+        # generate tf Variables in which we will store the model weights
+        self.model_stateholder = [tf.Variable(initial_value=w.read_value()) for w in self.model.trainable_weights]
+        self.partners_grads = [[tf.Variable(initial_value=w.read_value()) for w in self.model.trainable_weights]
+                               for _ in self.partners_list]
+        self.global_grad = [tf.Variable(initial_value=w.read_value()) for w in self.model.trainable_weights]
+        self.partners_optimizers = [self.model.optimizer.from_config(self.model.optimizer.get_config()) for _ in
+                                    self.partners_list]
+        if self.global_optimiser:
+            self.model.compile(optimizer=self.global_optimiser)
+
+    def fit(self):
+        # TF function definition
+        @tf.function
+        def fit_minibatch(model, model_stateholder, partners_minibatches, partners_optimizers, partners_grads,
+                          global_grad, aggregation_weights):
+            for model_w, old_w in zip(model.trainable_weights, model_stateholder):  # store model weights
+                old_w.assign(model_w.read_value())
+
+            for p_id, minibatch in enumerate(partners_minibatches):  # minibatch == (x,y)
+                # minibatch[0] in a tensor of shape=(number of batch, batch size, img).
+                # We cannot iterate on tensors, so we convert this tensor to a list of
+                # *number of batch* tensors with shape=(batch size, img)
+                x_minibatch = tf.unstack(minibatch[0], axis=0)
+                y_minibatch = tf.unstack(minibatch[1], axis=0)  # same here, with labels
+
+                for x, y in zip(x_minibatch, y_minibatch):  # iterate over batches
+                    with tf.GradientTape() as tape:
+                        y_pred = model(x)
+                        loss = model.compiled_loss(y, y_pred)
+                    model.compiled_metrics.update_state(y, y_pred)  # log the loss and accuracy
+                    partners_optimizers[p_id].minimize(loss, model.trainable_weights,
+                                                       tape=tape)  # perform local optimizations
+                # get the gradient as theta_before_minibatch - theta_after_minibatch
+                for grad_per_layer, w_old, w_new in zip(partners_grads[p_id], model_stateholder,
+                                                        model.trainable_weights):
+                    grad_per_layer.assign((w_old - w_new))
```

Comment:
@arthurPignet Here you're using weight differences between two consecutive minibatch updates as gradients, right? but how does that work on a theoretical level?

Reply:
Yes indeed. 
When you are performing a classic SGD update, the difference between the weights is the stocastic estimation of the gradient * lr. 
But here we are using more sophisticated optimization algorithms, with momentums and adaptative lr. The gradient is not easy to reach. So I fall back to an object which is not exactly the gradient, but the cumulative descent direction (which is the object of our interest actually). I compute the descent direction to be sent to the server as the vec. sum of the descents applied during the local batches. That corresponds to the difference between the weights at the beginning of the local batches and the ones at the ends of the local batches, just before the aggregation/global optimization which will conclude the global batch.  

Did I answer your question ? Maybe I misunderstood what you were saying.

Reply:
Yes thanks, it's easy to miss that the difference between the weights equates to gradient*lr, unless you have the SGD update rule in front of your eyes :). parhaps it would be much clearer for future readers to  to change the name of the parameter "global_grad" to something like cumulative_descent maybe, just like you explained. It's a bit misleading at first when reading the name 'grad' being used.

Reply:
Ok that is a nice idea. 
But I will use this cumulative_descent like a gradient just after for the aggregation and the global optimization. Don't you think that using a object which is not named grad where it is usually a gradient in every other algorithm, to play the role of the gradient could be even more misleading ?
Maybe rather than a reformating the variable I could add a comment saying that this thing is not really a gradient but a cumulative descent ?

Reply:
yes, a comment would be great.

Reply:
Or name it `pseudo_gradient` or something of that sort